src/main/java/ca/serum390/godzilla/api/handlers/ProductionManagerHandler.java

```diff
@@ -0,0 +1,253 @@
+                    int used_quantity = bomItem.getQuantity();
+                    if (plannedProduct.getUsedItems().containsKey(bomItemID)) {
+                        used_quantity += plannedProduct.getUsedItems().get(bomItemID);
+                    }
+                    plannedProduct.getUsedItems().put(bomItemID, used_quantity);
+                }
+                total_quantity = total_quantity - bomItem.getQuantity();
+                //update inventory
+                inventoryRepository.updateQuantity(bomItemID, 0).block();
+
+                // create purchase order this bom item with the number = total-quantity
+                if (purchaseOrder.getItems().containsKey(bomItemID)) {
+                    total_quantity += purchaseOrder.getItems().get(bomItemID);
+                }
+                purchaseOrder.getItems().put(bomItemID, total_quantity);
+            }
+        }
+    }
+
+    // if order is ready [all finished items in inventory] it sets the status of order to ready
+    // otherwise, it sets the status of production to processing, sets the status of production to scheduled or blocked
+    // saves the production item in db
+    private void processOrder() {
+        if (isOrderReady) {
+            ordersRepository.updateStatus(salesOrder.getId(), Order.READY).block();
+        } else {
+            ordersRepository.updateStatus(salesOrder.getId(), Order.PROCESSING).block();
+            plannedProduct.setStatus(isOrderBlocked ? PlannedProduct.BLOCKED : PlannedProduct.SCHEDULED);
+            PlannedProduct product = plannedProductsRepository.save(plannedProduct.getOrderID(), plannedProduct.getProductionDate(), plannedProduct.getStatus(), plannedProduct.getUsedItems()).block();
+            purchaseOrder.setProductionID(product.getId());
+
+            if (isOrderBlocked) {
+                schedulePurchaseOrder();
+            } else {
+                scheduleProduction(product.getId());
+            }
+        }
+    }
+
+
+    // saves the purchase order in db
+    //schedules the purchase order event to 2 minutes after current time
+    private void schedulePurchaseOrder() {
+        Order order = ordersRepository.save(purchaseOrder.getCreatedDate(), purchaseOrder.getDueDate(), purchaseOrder.getDeliveryLocation(), purchaseOrder.getOrderType(), purchaseOrder.getStatus(), purchaseOrder.getItems(), purchaseOrder.getProductionID()).block();
+        // schedule purchase order
+        PurchaseOrderEvent purchaseOrderEvent = new PurchaseOrderEvent(order.getId());
+        logger.info("purchase order " + order.getId() + " is created");
+        Runnable purchaseTask = () -> applicationEventPublisher.publishEvent(purchaseOrderEvent);
+        ScheduledExecutorService localExecutor = Executors.newSingleThreadScheduledExecutor();
+        scheduler = new ConcurrentTaskScheduler(localExecutor);
+        scheduler.schedule(purchaseTask, new Date(System.currentTimeMillis() + 120000));
```

Comment:
I think I understand the scheduling now:

1. We are registering some `@EventListener` methods for our custom events
2. Publishing events via the `ApplicationEventPublisher`
3. Scheduling the calls to the publisher via a `TaskScheduler` 

This will definitely be good enough for demo, but basically we are keeping this event system local to the app instance. The issue with that is that in the container cluster, our containers are "ephemeral" (i.e. disposable). The container orchestrator (AWS ECS) is free to destroy a container at any moment, and spin up fresh containers. 

Also, there may be more than one app instance running in the cluster. For the demo, we only have 1 container, but the app should be programmed as if there are multiple containers. You can never be sure which container you are connecting to when you browse to the app. This is not even considering the application crashing or something.

So for this project, probably this is good enough to demo, but we should just keep in mind as a goal to have stateless containers, or at the very least to persist all our state very often in the DB. It is harder to design these kinds of apps, but it makes them much more cloud-friendly and scalable when we can spin up as many containers as we want without having to worry about managing state inside the containers.

If we were working fulltime on this, probably it would be a good idea to completely remove all state from the app containers (even authentication) and to keep it all in external systems like: database (push state from app cluster into database cluster), external Message Queues (like RabbitMQ or Amazon has many services like this), external authentication services (custom services or cloud services like [Amazon Cognito](https://aws.amazon.com/cognito/)), and so on.

Like for this timed event system, we could develop either a custom service deployed independently from our app (and make this separate service stateful if need be), or investigate an off-the-shelf system that could be used as an event bus and subscribed to by all our app instances. Like maybe RabbitMQ could be used or Apache Kafka, or something like that that we could deploy to keep track of events for our application